hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/BlockContentsValidator.java

```diff
@@ -0,0 +1,123 @@
+                BlockStreamAccess.BLOCK_STREAM_ACCESS.readBlocks(node0Dir.resolve("data/block-streams/block-0.0.3"));
+        validator.validateBlocks(blocks);
+    }
+
+    public static final Factory FACTORY = new Factory() {
+        @NonNull
+        @Override
+        public BlockStreamValidator create(@NonNull final HapiSpec spec) {
+            return new BlockContentsValidator();
+        }
+    };
+
+    @Override
+    public void validateBlocks(@NonNull final List<Block> blocks) {
+        for (int i = 0; i < blocks.size(); i++) {
+            try {
+                validate(blocks.get(i), i == 0);
+            } catch (AssertionError err) {
+                logger.error("Error validating block {}", blocks.get(i));
+                throw err;
+            }
+        }
+    }
+
+    private static void validate(Block block, boolean isFirst) {
+        final var blockItems = block.items();
+
+        // A block SHALL start with a `header`.
+        if (!blockItems.getFirst().hasBlockHeader()) {
+            Assertions.fail("Block does not start with a block header");
+        }
+
+        // A block SHALL end with a `state_proof`.
+        if (!blockItems.getLast().hasBlockProof()) {
+            Assertions.fail("Block does not end with a block proof");
+        }
+
+        // For the first block the `block_header` might be followed by `event_header` or `state_changes`, due to state
+        // migration.
+        if (isFirst) {
+            if (!blockItems.get(1).hasEventHeader() && !blockItems.get(1).hasStateChanges()) {
+                Assertions.fail("First block header not followed by event header or state changes");
+            }
+            return;
+        }
+
+        // TODO: In general, a `block_header` SHALL be followed by an `event_header`, but for hapiTestRestart we get
+        // state change singleton update for BLOCK_INFO_VALUE
+        if (!blockItems.get(1).hasEventHeader() && !blockItems.get(1).hasStateChanges()) {
+            Assertions.fail("Block header not followed by an event header or state changes");
+        }
+
+        if (blockItems.stream().noneMatch(BlockItem::hasEventTransaction)) { // block without a user transaction
+            // A block with no user transactions contains a `block_header`, `event_headers`, `state_changes` and
+            // `state_proof`.
+            if (blockItems.stream()
+                    .skip(1)
+                    .limit(blockItems.size() - 2L)
+                    .anyMatch(item -> !item.hasEventHeader() && !item.hasStateChanges())) {
+                Assertions.fail(
+                        "Block with no user transactions should contain items of type `block_header`, `event_headers`, `state_changes` or `state_proof`");
+            }
+
+            return;
+        }
+
+        for (int i = 0; i < blockItems.size(); i++) {
+            // TODO: An `event_header` SHALL be followed by one or more `event_transaction` items. -> looks like we can
```

Comment:
Hmm yes I think it's very easy for there to be an `EventHeader` with no following `EventTransaction` because not every round will include a user transaction.

Seems fine to me.

Reply:
Removed the comment with the commented-out code since we'll not be enforcing the `event_header SHALL be followed by one or more event_transaction` rule.